File: reorder_list.py
# https://leetcode.com/problems/reorder-list/description/
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
    def __repr__(self) -> str:
        return str(self.val)

class Solution:
    def reorderList(self, head) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        slow = head
        while slow:
            prev = slow
            fast = slow.next
            while fast and fast.next: # Find tail
                prev = fast
                fast = fast.next
                if fast.next == slow: # Only can happen on cycle
                    fast.next = None

            prev.next = None
            fast.next = slow.next
            slow.next = fast

            slow = fast.next # We don't want to reorder the one we just swapped

        return head
    
def main():
    s = Solution()
    print(s.reorderList(ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5)))))))

if __name__ == '__main__':
    main()

File: reorder_list.java
/**
 * https://leetcode.com/problems/reorder-list/
 */
import java.util.Stack;
import java.lang.Math;

public class ListNode {
     int val;
     ListNode next;
     ListNode() {}
     ListNode(int val) { this.val = val; }
     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 }

class Solution {
    public ListNode solve(ListNode head) {
        ListNode slow = head;
        ListNode fast =  head.next;

        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

        // Reverse from slow to end
        ListNode prev = null;
        ListNode next = slow.next;
        slow.next = null; // This wil be tail

        while (next != null) {
            ListNode tmp = next.next;
            next.next = prev;
            prev = next;
            next = tmp;
        }

        // Merge
        ListNode node = head;

        while (prev != null) {
            ListNode tmp = node.next, tmp2 = prev.next;
            node.next = prev;
            prev.next = tmp;
            node = tmp;
            prev = tmp2;
        }
    }
}
public class Main {
    public static void main(String[] args) {
        Solution s = new Solution();
        System.out.println(s.solve());
    }
}